OpenRA.Mods.Common/Widgets/ObserverProductionIconsWidget.cs

```diff
@@ -77,7 +96,9 @@ public override void Draw()
 				if (!clocks.ContainsKey(queue.Trait))
 					clocks.Add(queue.Trait, new Animation(world, ClockAnimation));
 
+			icons = new Dictionary<Rectangle, ProductionIcon>();
```

Comment:
Throwing out and regenerating this 60 times per second will churn memory unnecessarily.

You know at widget creation time the maximum number of icons that will be drawn, so you could split this into a `Rectangle[]` (set up once and then left alone) and a `ProductionIcon[]` that you update by index each frame here.  The LINQ in `HandleMouseInput` (&rarr; `Tick`) can then be replaced with a `for` loop so that you can use the index of the matching `Rectangle` to get the corresponding `ProductionIcon`.